### ISSUE TITLE
Question about CONCH feature extraction in HiVE-MIL

### ISSUE THREAD

**Issue by HuZichun1** [bryanwong17/HiVE-MIL#4]
Hi, thanks for your great work!

I am trying to reproduce HiVE-MIL and I have matched the following components:

- Same dataset(TCGA-BRCA) and splits
- Same patch coordinates (5x / 20x)
- Same hierarchical structure (1→16 mapping verified)
- Same training code and hyperparameters

However, I still observe a noticeable performance drop (especially in some folds).

After careful debugging, I suspect the difference may come from the feature extraction pipeline.

Could you please clarify the following?

1. **CONCH weights**: Which Hugging Face repository / revision was used for the released features?
  - `MahmoodLab/conch` (original), or
  - `MahmoodLab/conchv1_5` (CONCH v1.5), or another checkpoint?
  If possible, the exact HF repo id + commit / file revision would help a lot.
2. **Feature extraction settings**: For CONCH, did you use the same setting as in this repo’s `feature_extraction/patch_extraction_utils.py`, i.e. `encode_image(..., proj_contrast=True, normalize=True)` with the provided eval transforms?
3. **Hierarchical features**: When building `hierarchical_*` from 5× and 20× features, was `np.random.choice` for subsampling (when a region has more than `max_patches` high-mag patches) run with a **fixed random seed**? If yes, what seed (or was it only run once and the release is from that single run)?
4. **Optional**: If the released features were produced with a slightly different internal script than the current public repo, any pointer to the exact script version or commit would be appreciated.
Thank you again for your time and for open-so

**Comment by bryanwong17** [comment]
Hi @HuZichun1,

Thanks for your interest in HiVE-MIL!

### 1) CONCH weights
I used the original CONCH weights with an embedding dimension of 512. The model can be loaded using the following code:

```python
model_name = "conch_ViT-B-16"
model_path = "hf_hub:MahmoodLab/conch"
hf_auth_token = ""  # Add your Hugging Face auth token here

model, eval_t = create_model_from_pretrained(
    model_name,
    model_path,
    hf_auth_token=hf_auth_token
)
```

### 2) Feature extraction settings
Yes, the image embeddings were obtained using the following code:

```python
embeds = model.encode_image(batch, proj_contrast=True, normalize=True).detach().cpu().numpy().squeeze()
```

### 3) Hierarchical features
Unfortunately, I did not set a fixed random seed for this experiment, and it was run only once.


If you still encounter difficulties reproducing the results, I recommend downloading the preprocessed WSI features provided in the README under: `4.1 Download Preprocessed Hierarchical WSI Features (Recommended)`

Hope this helps!

**Comment by HuZichun1** [comment]
Thanks a lot for your reply, I really appreciate it!

By the way, I’m currently working on CAMELYON16 as well. 
Would you mind sharing the prompts and feature you used, if possible? 
Thank you again!

**Comment by bryanwong17** [comment]
Hi @HuZichun1, apologies for the late reply!

It turns out I no longer have the exact prompts or features I used for CAMELYON16. Therefore, I'd recommend to extract CONCH features and build the hierarchical features using the provided code.

For the prompt template, you can refer to Section 3.1 of the main paper.

Thanks again for your interest in HiVE-MIL!

**Comment by HuZichun1** [comment]
Thank you so much!

**Comment by bryanwong17** [comment]
You're welcome! If you find the repo helpful, I’d really appreciate it if you could give it a star 🌟 and share it with others who might find it useful!
